pkg/storage/stores/tsdb/cached_postings_index.go

```diff
@@ -0,0 +1,186 @@
+package tsdb
+
+import (
+	"context"
+	"sort"
+	"strings"
+
+	"github.com/dennwc/varint"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	promEncoding "github.com/prometheus/prometheus/tsdb/encoding"
+
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/storage/stores/tsdb/index"
+	"github.com/grafana/loki/pkg/util/encoding"
+)
+
+type PostingsClient interface {
+	ForPostings(ctx context.Context, matchers []*labels.Matcher, fn func(index.Postings) error) error
+}
+
+var sharedCacheClient cache.Cache
+
+func NewCachedPostingsClient(reader IndexReader, logger log.Logger, cacheClient cache.Cache) PostingsClient {
+	return &cachedPostingsClient{
+		reader:      reader,
+		cacheClient: cacheClient,
+		log:         logger,
+	}
+}
+
+type cachedPostingsClient struct {
+	reader IndexReader
+
+	cacheClient cache.Cache
+
+	log log.Logger
+}
+
+func (c *cachedPostingsClient) ForPostings(ctx context.Context, matchers []*labels.Matcher, fn func(index.Postings) error) error {
+	key := CanonicalLabelMatchersKey(matchers)
+	if postings, got := c.fetchPostings(key); got {
+		return fn(postings)
+	}
+
+	p, err := PostingsForMatchers(c.reader, nil, matchers...)
+	if err != nil {
+		return err
+	}
+
+	if err := c.storePostings(p, matchers); err != nil {
+		level.Error(c.log).Log("msg", "failed to cache postings", "err", err, "matchers", key)
+	}
+	return fn(p)
+}
+
+// diffVarintEncodeNoHeader encodes postings into diff+varint representation.
+// It doesn't add any header to the output bytes.
+// Length argument is expected number of postings, used for preallocating buffer.
+func diffVarintEncodeNoHeader(p index.Postings, length int) ([]byte, error) {
+	buf := encoding.Encbuf{}
+
+	// This encoding uses around ~1 bytes per posting, but let's use
+	// conservative 1.25 bytes per posting to avoid extra allocations.
+	if length > 0 {
+		buf.B = make([]byte, 0, 5*length/4)
+	}
+
+	prev := storage.SeriesRef(0)
+	for p.Next() {
+		v := p.At()
+
+		// TODO(dylanguedes): can we ignore this?
+		// if v < prev {
+		// 	return nil, errors.Errorf("postings entries must be in increasing order, current: %d, previous: %d", v, prev)
+		// }
+
+		// This is the 'diff' part -- compute difference from previous value.
+		buf.PutUvarint64(uint64(v - prev))
+		prev = v
+	}
+	if p.Err() != nil {
+		return nil, p.Err()
+	}
+
+	return buf.B, nil
+}
+
+func decodeToPostings(b []byte) index.Postings {
+	decoder := encoding.DecWrap(promEncoding.Decbuf{B: b})
+
+	refs := []storage.SeriesRef{}
```

Comment:
This is kinda expensive if we don't know the size of the refs upfront. Maybe using a pool?

Reply:
I found a workaround doing the following:
- When encoding to store on the cache, first add the number of seriesRefs
- When decoding, fetch the first uint64, it should be the number of series in the postings. This way we can alloc the right number of refs straightaway